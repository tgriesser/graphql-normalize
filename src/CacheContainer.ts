import produce, { Immutable } from 'immer';
import type { CacheShape } from './cache';

export class CacheContainer {
  #state: Immutable<CacheShape>;
  #subscribers: Function[] = [];

  constructor(initialState?: CacheShape) {
    this.#state = produce({ fields: {}, operations: {} }, (root) => initialState ?? root);
  }

  getState() {
    return this.#state;
  }

  subscribe(fn: (state: Immutable<CacheShape>) => void) {
    this.#subscribers.push(fn);
    return () => {
      this.#subscribers.splice(this.#subscribers.indexOf(fn), 1);
    };
  }
}
