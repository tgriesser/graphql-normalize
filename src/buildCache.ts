import type { FormattedExecutionResult } from 'graphql';
import type { ArgDef, FieldDef, FieldMeta, NormalizedDoc } from './extensionShape';
import { stringifyVariables } from './stringifyVariables';

interface BuildCacheObj {
  meta: NormalizedDoc;
  result: FormattedExecutionResult;
  variableValues: Record<string, any>;
}

const NO_ARGS = '^NO_ARGS';

/**
 * Write values into the cache, given the values, the existing cache,
 * and the new values that need to be written
 */
export function buildCache(obj: BuildCacheObj) {
  const { meta, result, variableValues } = obj;

  const normalizedData: Record<string, any> = {};

  const currentDataStack: any[] = [];
  let currentData: any = result.data ?? {};

  const currentCacheStack: any[] = [];
  let currentCache = normalizedData;

  function setCurrentData(data: any) {
    currentDataStack.push(currentData);
    currentData = data;
  }
  function popCurrentData() {
    currentData = currentDataStack.pop();
  }
  function setCurrentCache(obj: any) {
    currentCacheStack.push(currentCache);
    currentCache = obj;
  }
  function popCurrentCache() {
    currentCache = currentCacheStack.pop();
  }

  function getValue(key: string) {
    return currentData[key];
  }

  function getArgValue(key: string) {
    return variableValues[key] ?? meta.variables.find((k) => k.name === key)?.defaultValue;
  }

  function printArgs(args: string | ArgDef | undefined) {
    if (args === undefined) return NO_ARGS;
    if (typeof args === 'string') return args;
    const argsObj: Record<string, any> = {};
    for (const key of Object.keys(args).sort()) {
      if (key.startsWith('$')) {
        argsObj[args[key]] = getArgValue(key.slice(1));
      }
    }
    return stringifyVariables(argsObj);
  }

  // We want to iterate the known structure of the response shape, setting the values
  // in the result cache
  function iterateFields(fields: FieldDef[]) {
    for (const field of fields) {
      if (field === '__typename') {
        currentCache[field] = getValue(field);
      } else if (typeof field === 'string') {
        currentCache[field] ??= {};
        currentCache[field]['^NO_ARGS'] = getValue(field);
      } else {
        _writeField(field);
      }
    }
  }

  function _writeField(field: FieldMeta, keyPath?: string[]) {
    setCurrentData(currentData[field.alias ?? field.name]);
    currentCache[field.name] ??= {};

    const fieldArgs = printArgs(field.args);

    keyPath ??= [field.name, fieldArgs];

    currentCache[field.name][fieldArgs] ??= getContainer(field);

    // When we have a list of items, we want to map over those, and treat
    // them as the metadata tells us to
    if (field.list) {
      if (currentData === null) {
        currentCache[field.name][fieldArgs] = null;
      } else {
        _handleList(field, fieldArgs);
      }
    } else {
      if (field.cacheKey) {
        const $ref = makeCacheKey(field.cacheKey, currentData);
        normalizedData[$ref] ??= {};
        currentCache[field.name][fieldArgs] = { $ref };
        setCurrentCache(normalizedData[$ref]);
      } else {
        setCurrentCache(currentCache[field.name][fieldArgs]);
      }

      if (field.possible) {
        // console.log(field, currentData);
      } else if (field.fields) {
        iterateFields(field.fields);
      }

      popCurrentCache();
    }

    popCurrentData();
  }

  function _handleList(field: FieldMeta, fieldArgs: string) {
    const { list, ...rest } = field;
    const listItem = (currentCache[field.name][fieldArgs] ??= []);

    for (const [idx, value] of (currentData as Array<any>).entries()) {
      if (value === null) {
        listItem.push(value);
        continue;
      }

      setCurrentData(value);
      if (field.cacheKey) {
        const $ref = makeCacheKey(field.cacheKey, value);
        listItem.push({ $ref });
        normalizedData[$ref] ??= {};
        setCurrentCache(normalizedData[$ref]);
        if (field.fields) {
          iterateFields(field.fields);
        }
        popCurrentCache();
      }

      if (field.possible) {
        if (field.possible[currentData.__typename]) {
          const unionMeta = field.possible[currentData.__typename];
          if (unionMeta.cacheKey) {
            const $ref = makeCacheKey(unionMeta.cacheKey, value);
            listItem.push({ $ref });
            normalizedData[$ref] ??= {};
            setCurrentCache(normalizedData[$ref]);
            if (unionMeta.fields) {
              iterateFields(unionMeta.fields);
            }
            popCurrentCache();
          }
        } else {
          listItem.push(currentData);
        }
      }

      popCurrentData();
    }
  }

  iterateFields(meta.fields);

  return normalizedData;
}

function makeCacheKey(key: string, obj: any) {
  const [type, fields] = key.split(':');
  if (fields.indexOf(',') !== -1) {
    return `${type}:${stringifyVariables(fields.split(',').map((k) => obj[k]))}`;
  }
  return `${type}:${obj[fields]}`;
}

function getContainer(field: FieldMeta) {
  return field.list ? [] : {};
}
