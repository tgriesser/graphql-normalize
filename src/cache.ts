import type { NormalizeMetaShape } from './metadataShapes';

export interface CacheShape {
  operations: {
    // The metadata for each query
    [hash: string]: {
      meta: NormalizeMetaShape;
    };
  };
  fields: {
    [key: string]: any;
  };
}
