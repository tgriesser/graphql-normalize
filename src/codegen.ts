import { NormalizeCodegenPlugin } from './codegen/plugin.js';
export type { TypePolicies } from './codegen/getCacheKey.js';
export type { NormalizeMetaShape, NormalizedDocShape } from './metadataShapes.js';

export { NormalizeCodegenPlugin } from './codegen/plugin.js';
export { generateNormalizedMetadataForDocs, generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperations, generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

export { NormalizeCodegenPlugin as plugin };

export default {
  plugin: NormalizeCodegenPlugin,
};
