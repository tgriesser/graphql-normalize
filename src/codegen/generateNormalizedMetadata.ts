import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer'
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  OperationTypeNode,
  isObjectType,
  isScalarType,
  InlineFragmentNode,
  getNamedType,
  assertObjectType,
  isEnumType,
} from 'graphql'
import type {
  FieldDef,
  FieldMeta,
  NormalizeMetaShape,
  NormalizedDocShape,
  UnionMeta,
  VariableMeta,
} from '../metadataShapes'
import { TypePolicies, getCacheKey } from './getCacheKey'
import { stringifyVariables } from '../stringifyVariables'

export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedMetadataForDocs(schema, [document], typePolicies)[0].meta
}

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadataForDocs(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
): NormalizedDocShape[] {
  let finalDocuments = documents
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    finalDocuments = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    })
  }

  const normalizedDocs: NormalizedDocShape[] = []

  const typeInfo = new TypeInfo(schema)
  let normalizedDoc: NormalizeMetaShape = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    fields: [],
  }

  let parentStack: Array<FieldMeta | UnionMeta | NormalizeMetaShape> = []
  let parentFieldDef: FieldMeta | NormalizeMetaShape | UnionMeta = normalizedDoc

  function popStack() {
    const parent = parentStack.pop()
    if (parent) {
      parentFieldDef = parent
    }
  }

  function pushField(field: string | FieldDef) {
    parentFieldDef.fields ??= []
    if (typeof field === 'string' && parentFieldDef.fields.includes(field)) {
      return
    }
    parentFieldDef.fields.push(field)
    if (typeof field !== 'string') {
      parentStack.push(parentFieldDef)
      parentFieldDef = field
    }
  }

  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false
    const argsDef: Record<string, any> = {}
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.value.name.value}`] = arg.name.value
        hasVar = true
      } else {
        argsDef[arg.name.value] = print(arg.value)
      }
    }
    return hasVar ? argsDef : stringifyVariables(argsDef)
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition: {
      enter(op) {
        normalizedDoc.operation = op.operation
      },
      leave(op) {
        normalizedDocs.push({
          name: op.name?.value ?? 'unnamed',
          meta: normalizedDoc,
        })
        normalizedDoc = {
          operation: OperationTypeNode.QUERY,
          variables: [],
          fields: [],
        }
        parentFieldDef = normalizedDoc
        parentStack = []
      },
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      }
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue)
      }
      normalizedDoc.variables.push(obj)
    },
    Field: {
      enter(node) {
        const { gqlType, listDepth } = unpackType(typeInfo)

        // If this is a boring ol scalar / enum, we push it as a string
        if (isSimpleField(node, typeInfo)) {
          pushField(node.name.value)
          return
        }

        const fieldMeta: FieldMeta = { name: node.name.value }
        pushField(fieldMeta)

        if (listDepth) fieldMeta.list = listDepth
        if (node.alias?.value) fieldMeta.alias = node.alias.value

        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType)
          if (cacheKey) {
            fieldMeta.cacheKey = cacheKey
          }
          if (gqlType.name === schema.getQueryType()?.name) {
            fieldMeta.cacheKey = '$root'
          }
        }
        if (node.arguments?.length) {
          fieldMeta.args = normalizeArgs(node.arguments)
        }

        const skipDirective = node.directives?.find((d) => d.name.value === 'skip')
        const includeDirective = node.directives?.find((d) => d.name.value === 'include')
        if (skipDirective) {
          fieldMeta.skip = normalizeArgs(skipDirective.arguments ?? [])
        }
        if (includeDirective) {
          fieldMeta.include = normalizeArgs(includeDirective.arguments ?? [])
        }
      },
      leave(node) {
        // Pop things back into place
        if (!isSimpleField(node, typeInfo)) {
          popStack()
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (
          node.typeCondition?.name.value &&
          isConcreteAbstract(node, typeInfo, schema) &&
          parentIsFieldMeta(parentFieldDef)
        ) {
          const unionType: UnionMeta = {
            fields: [],
          }
          const outputType = getNamedType(typeInfo.getType())
          const cacheKey = getCacheKey(typePolicies, assertObjectType(outputType))
          if (cacheKey) {
            unionType.cacheKey = cacheKey
          }

          parentFieldDef.possible ??= {}
          parentFieldDef.possible[node.typeCondition?.name.value] = unionType
          parentStack.push(parentFieldDef)
          parentFieldDef = unionType
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value && isConcreteAbstract(node, typeInfo, schema)) {
          popStack()
        }
      },
    },
    FragmentDefinition() {
      throw noFragments()
    },
  })

  for (const doc of finalDocuments) {
    visit(doc, visitor)
  }

  return normalizedDocs
}

function unpackType(typeInfo: TypeInfo) {
  let listDepth = 0
  // We want to unwrap any list types, and make those the containing
  let gqlType = typeInfo.getType()
  while (isWrappingType(gqlType)) {
    if (isListType(gqlType)) {
      ++listDepth
    }
    gqlType = gqlType.ofType
  }
  return { gqlType, listDepth }
}

function isSimpleField(node: FieldNode, typeInfo: TypeInfo) {
  const { gqlType, listDepth } = unpackType(typeInfo)
  return (
    (isScalarType(gqlType) || isEnumType(gqlType)) &&
    !listDepth &&
    !node.alias &&
    !node.arguments?.length &&
    !node.directives?.some((d) => d.name.value === 'skip' || d.name.value === 'includes')
  )
}

function noFragments() {
  return new Error('Fragment definitions have been stripped')
}

function isConcreteAbstract(node: InlineFragmentNode, typeInfo: TypeInfo, schema: GraphQLSchema) {
  if (node.typeCondition?.name) {
    const parentTypeName = typeInfo.getParentType()?.name
    if (node.typeCondition.name.value !== parentTypeName) {
      return isObjectType(schema.getType(node.typeCondition?.name.value))
    }
  }
  return false
}

function parentIsFieldMeta(parent: FieldMeta | NormalizeMetaShape | UnionMeta): parent is FieldMeta {
  return 'name' in parent
}
