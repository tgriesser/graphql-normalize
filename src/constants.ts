export const __typename = '__typename' as const;

export const NO_ARGS = '$';
