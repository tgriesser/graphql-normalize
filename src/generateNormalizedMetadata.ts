import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isAbstractType,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  SelectionSetNode,
  OperationTypeNode,
  isObjectType,
} from 'graphql';
import stableStringify from 'fast-json-stable-stringify';
import type { FieldDef, FieldMeta, NormalizeMeta, NormalizedDoc, VariableMeta } from './extensionShape';
import { TypePolicies, getCacheKey } from './typePolicies';

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  operation: DocumentNode,
  typePolicies: TypePolicies = {}
): NormalizedDoc {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }

  const typeInfo = new TypeInfo(schema);
  const normalizedDoc: NormalizedDoc = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    selectionSet: {},
  };

  const fieldPath: any[] = [];

  function setPath<K extends keyof NormalizeMeta>(key: K, value: NormalizeMeta[K]) {
    normalizedDoc.selectionSet[fieldPath.join('.')] ??= {};
    normalizedDoc.selectionSet[fieldPath.join('.')][key] = value;
  }
  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false;
    const argsDef: Record<string, any> = {};
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.name.value}`] = arg.name.value;
        hasVar = true;
      } else {
        argsDef[arg.name.value] = print(arg.value);
      }
    }
    return hasVar ? argsDef : stableStringify(argsDef);
  }
  function normalizeField(node: FieldNode) {
    if (node.alias || node.arguments?.length) {
      const field: FieldMeta = {
        name: node.name.value,
      };
      if (node.alias) field.alias = node.alias.value;
      if (node.arguments?.length) field.args = normalizeArgs(node.arguments);
      return field;
    }
    return node.name.value;
  }
  function addSelectionFields(node: SelectionSetNode) {
    const fields: FieldDef[] = [];
    for (const sel of node.selections) {
      if (sel.kind === Kind.FRAGMENT_SPREAD) {
        throw noFragments();
      }
      if (sel.kind === Kind.FIELD) {
        fields.push(normalizeField(sel));
      }
      // If we have an inline fragment
      if (sel.kind === Kind.INLINE_FRAGMENT) {
        //
      }
    }
    setPath('fields', fields);
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition(op) {
      normalizedDoc.operation = op.operation;
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      };
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue);
      }
      normalizedDoc.variables.push(obj);
    },
    Field: {
      enter(node) {
        if (node.alias?.value) {
          fieldPath.push(node.alias.value);
          setPath('aliasField', node.name.value);
        } else {
          fieldPath.push(node.name.value);
        }
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            setPath('list', true);
            fieldPath.push('$idx');
          }
          gqlType = gqlType.ofType;
        }
        if (isAbstractType(gqlType)) {
          setPath(
            'possible',
            schema.getPossibleTypes(gqlType).map((t) => t.name)
          );
        }
        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType);
          if (cacheKey) {
            setPath('cacheKey', cacheKey);
          }
        }
        if (node.arguments?.length) {
          setPath('args', normalizeArgs(node.arguments));
        }
        if (node.selectionSet) {
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        fieldPath.pop();
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            fieldPath.pop();
          }
          gqlType = gqlType.ofType;
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (node.typeCondition?.name) {
          fieldPath.push(`$type_${node.typeCondition?.name.value}`);
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value) {
          fieldPath.pop();
        }
      },
    },
    FragmentDefinition() {
      throw noFragments();
    },
  });

  visit(finalOperation, visitor);

  return normalizedDoc;
}

function noFragments() {
  return new Error('Fragment definitions have been stripped');
}
