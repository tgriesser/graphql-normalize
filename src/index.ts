import { graphqlNormalize } from './graphqlNormalize.js';
export { stringifyVariables } from './stringifyVariables.js';
export { MissingFieldError } from './errors.js';
export type { graphqlNormalizeOptions, NormalizeMetaShape } from './graphqlNormalize.js';
export { graphqlNormalize };
export default graphqlNormalize;
