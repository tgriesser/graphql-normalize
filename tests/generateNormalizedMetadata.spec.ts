import { describe, expect, it } from 'vitest';
import { generateNormalizedMetadata } from '../src/generateNormalizedMetadata';
import { schema } from './fixtures/schema';
import { operation1Doc } from './fixtures/ops';

describe('generateNormalizedMetadata', () => {
  //
  it('should generate the necessary metadata', () => {
    expect(generateNormalizedMetadata(schema, operation1Doc)).toMatchSnapshot();
  });

  //
});
