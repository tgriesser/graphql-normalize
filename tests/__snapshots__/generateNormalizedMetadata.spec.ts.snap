// Vitest Snapshot v1

exports[`generateNormalizedMetadata > should generate the necessary metadata 1`] = `
{
  "fields": [
    {
      "args": {
        "$first": "first",
      },
      "fields": [
        {
          "cacheKey": "Post:id",
          "fields": [
            "id",
            {
              "cacheKey": "User:id",
              "fields": [
                "id",
              ],
              "name": "author",
            },
          ],
          "list": 1,
          "name": "nodes",
        },
        {
          "cacheKey": "QueryPostsConnectionEdge:cursor",
          "fields": [
            {
              "cacheKey": "Post:id",
              "fields": [
                "id",
                {
                  "cacheKey": "User:id",
                  "fields": [
                    "id",
                  ],
                  "name": "author",
                },
              ],
              "name": "node",
            },
          ],
          "list": 1,
          "name": "edges",
        },
      ],
      "name": "posts",
    },
    {
      "alias": "first10Posts",
      "args": "{\\"first\\":\\"10\\"}",
      "fields": [
        {
          "cacheKey": "Post:id",
          "fields": [
            "id",
          ],
          "list": 1,
          "name": "nodes",
        },
      ],
      "name": "posts",
    },
    {
      "alias": "currentNode",
      "args": {
        "$nodeId": "id",
      },
      "name": "node",
      "possible": {
        "Post": {
          "cacheKey": "Post:id",
          "fields": [
            "id",
          ],
        },
        "User": {
          "cacheKey": "User:id",
          "fields": [
            "__typename",
            "name",
          ],
        },
      },
    },
    {
      "args": "{\\"first\\":\\"10\\"}",
      "fields": [
        {
          "list": 1,
          "name": "nodes",
          "possible": {
            "Blog": {
              "cacheKey": "Blog:id",
              "fields": [
                "id",
                "name",
              ],
            },
            "Post": {
              "cacheKey": "Post:id",
              "fields": [
                "__typename",
                "id",
              ],
            },
          },
        },
      ],
      "name": "homeItems",
    },
    "num",
  ],
  "operation": "query",
  "variables": [
    {
      "defaultValue": "12",
      "name": "first",
    },
    {
      "defaultValue": "\\"\\"",
      "name": "nodeId",
    },
    {
      "defaultValue": "false",
      "name": "hasNode",
    },
  ],
}
`;
